internal/jujuapi/streamproxy.go

```diff
@@ -0,0 +1,150 @@
+		var errResult jujuparams.ErrorResult
+		errResult.Error = &jujuparams.Error{
+			Message: msg,
+			Code:    string(code),
+		}
+		err := clientConn.WriteJSON(errResult)
+		if err != nil {
+			zapctx.Error(ctx, "failed to write error message to client", zap.Error(err), zap.Any("client message", errResult))
+		}
+	}
+	user, err := s.jimm.UserLogin(ctx, auth.SessionIdentityFromContext(ctx))
+	if err != nil {
+		zapctx.Error(ctx, "user login error", zap.Error(err))
+		writeError(err.Error(), errors.CodeUnauthorized)
+		return
+	}
+	uuid, finalPath, err := modelInfoFromPath(jimmhttp.PathElementFromContext(ctx, "path"))
+	if err != nil {
+		zapctx.Error(ctx, "error parsing path", zap.Error(err))
+		writeError(fmt.Sprintf("error parsing path: %s", err.Error()), errors.CodeBadRequest)
+		return
+	}
+	model, err := s.getModel(ctx, uuid)
+	if err != nil {
+		writeError(err.Error(), errors.CodeModelNotFound)
+		return
+	}
+	if ok, err := checkPermission(ctx, finalPath, user, model.ResourceTag()); err != nil {
+		writeError(err.Error(), errors.CodeUnauthorized)
+		return
+	} else if !ok {
+		writeError(fmt.Sprintf("unauthorized access to endpoint: %s", finalPath), errors.CodeUnauthorized)
+		return
+	}
+	api, err := s.jimm.Dialer.Dial(ctx, &model.Controller, model.ResourceTag(), nil)
+	if err != nil {
+		zapctx.Error(ctx, "failed to dial controller", zap.Error(err))
+		writeError(fmt.Sprintf("failed to dial controller: %s", err.Error()), errors.CodeConnectionFailed)
+		return
+	}
+	defer api.Close()
+	controllerStream, err := api.ConnectStream(finalPath, nil)
+	if err != nil {
+		zapctx.Error(ctx, "failed to connect stream", zap.Error(err))
+		writeError(fmt.Sprintf("failed to connect stream: %s", err.Error()), errors.CodeConnectionFailed)
+		return
+	}
+	proxyStreams(clientConn, controllerStream)
+}
+
+func (s streamProxier) getModel(ctx context.Context, modelUUID string) (dbmodel.Model, error) {
+	model := dbmodel.Model{
+		UUID: sql.NullString{
+			String: modelUUID,
+			Valid:  modelUUID != "",
+		},
+	}
+	if err := s.jimm.Database.GetModel(context.Background(), &model); err != nil {
+		zapctx.Error(ctx, "failed to find model", zap.String("uuid", modelUUID), zap.Error(err))
+		return dbmodel.Model{}, fmt.Errorf("failed to find model: %s", err.Error())
+	}
+	return model, nil
+}
+
+// proxyStreams starts a simple proxy for 2 websockets.
+// After starting the proxy we listen for the first error
+// returned and then close both connections before waiting
+// for an error from the second connection.
+func proxyStreams(src, dst base.Stream) {
+	errChan := make(chan error, 2)
+	go func() { errChan <- proxy(src, dst) }()
+	go func() { errChan <- proxy(dst, src) }()
+	<-errChan
+	dst.Close()
+	src.Close()
+	<-errChan
+}
+
+func proxy(src base.Stream, dst base.Stream) error {
+	for {
+		var data map[string]any
+		err := src.ReadJSON(&data)
+		if err != nil {
+			return err
+		}
+		err = dst.WriteJSON(data)
+		if err != nil {
```

Comment:
i think you should use the same approach you used in a previous pr, where you don't return errors for closed connections, but you do for json errors.

Reply:
I would agree, but I don't think it's really possible for the marshal/unmarshal parts to trigger an error in this scenario. But I'll take another look at this.

Reply:
Moved the stream proxy to the rpc package where it feels more appropriate and now I can reuse that work.